### PR TITLE
[FEATURE][ML] Boosted regression tree (initial commit)

### PR DESCRIPTION
**Pull request by tveasey**
This implements a boosted regression tree. This is still very much work in progress and largely untested, but it is a first pass implementation of nearly the full train functionality and defines the API for other tasks. As such, I think this represents a point where we can review without the change becoming too large.

The biggest missing piece is hyperparameter optimisation. This will build on some standalone functionality, so I'm leaving it for the time being. I also have some plans for custom treatment of categorical regressors, but this depends on adding a column data type and I'm going to delay this until later.

### CODE REVIEW

File: lib/maths/CBoostedTree.cc [hunk 1]
```diff
@@ -0,0 +1,1107 @@
...
+        // TODO Hyperparameter optimisation...
+        // Get next parameters to test, compute generalisation error, record if best, repeat.
+
+        LOG_TRACE(<< "Starting training");
+
+        TMeanAccumulator meanLoss;
+        for (std::size_t i = 0; i < m_NumberFolds; ++i) {
+            LOG_TRACE(<< "fold = " << i);
+            TNodeVecVec forest(this->trainForest(frame, trainingRowMasks[i], recordProgress));
+            double loss{this->meanLoss(frame, testingRowMasks[i], forest)};
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTree.cc]: I debated this. I think since it is on a hold out set there is no need to additionally penalise model complexity since if it overfit it should generalise badly and the error will be high. It certainly isn't clear cut to me. We can experiment with this down the line though.

File: lib/maths/CBoostedTree.cc [hunk 2]
```diff
@@ -0,0 +1,1107 @@
...
+            candidateSplits, this->featureBag(frame), trainingRowMask));
+
+        double totalGain{0.0};
+
+        for (std::size_t i = 0; i < maximumTreeSize; ++i) {
+
+            auto split = leaves.top();
+            leaves.pop();
+
+            if (split->gain() < 1e-7 * totalGain) {
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTree.cc]: ++

File: lib/maths/CBoostedTree.cc [hunk 3]
```diff
@@ -0,0 +1,1107 @@
...
+                break;
+            }
+
+            totalGain += split->gain();
+            LOG_TRACE(<< "total gain = " << totalGain);
+
+            std::size_t splitFeature;
+            double splitValue;
+            std::tie(splitFeature, splitValue) = split->bestSplit();
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTree.cc]: Agreed

File: lib/maths/CBoostedTree.cc [hunk 4]
```diff
@@ -0,0 +1,1107 @@
...
+CMse::TArgMinLossUPtr CMse::minimizer() const {
+    return std::make_unique<CArgMinMse>();
+}
+
+double CArgMinMse::value() const {
+    return CBasicStatistics::mean(m_MeanError);
+}
+}
+
+class CBoostedTree::CImpl final {
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTree.cc]: So I have some protection for no progress [here](https://github.com/elastic/ml-cpp/pull/443/files#diff-4562017aa975b37520b5ea56d78e274cR865), although this still needs proper evaluation. The forest size also limits this, as does the mechanism we use for choosing initial values of regularisation parameters. However, I really see this as tightly coupled with the whole hyperparameter optimisation loop, which I'm going to work on soon. I'll leave a TODO in the code to make sure we have proper sanity checking though.

**Comment by tveasey** [comment]
Regarding bit masks, yes we do loose some locality of reference, but I feel it makes the implementation very clean and is efficient if everything is in core: no copying, efficient iteration and some operations can be done directly on the bit masks and don't need to visit the data at all. Also, if everything is in core, then provided we repeatedly work on similar nodes, we should have the relevant rows in, at least, one of the caches.

In general, there might be a case for maintaining some sort of cache on top of the data frame, but this complicates the implementation and we'd have to control the memory usage of this somehow.

In some initial testing of the current implementation it seems pretty performant. I don't want to optimise for this early and I don't want to make any changes without being able to test the final performance including hyperparameter optimisation.

**Comment by tveasey** [comment]
Thanks for the review @benwtrent.
